distribution/pull.go

```diff
@@ -115,36 +115,32 @@ func Pull(ctx context.Context, ref reference.Named, imagePullConfig *ImagePullCo
 		if err := puller.Pull(ctx, ref); err != nil {
 			// Was this pull cancelled? If so, don't try to fall
 			// back.
-			fallback := false
 			select {
 			case <-ctx.Done():
+				logrus.Errorf("Not continuing with pull after error: %v", err)
+				return TranslatePullError(err, ref)
 			default:
-				if fallbackErr, ok := err.(fallbackError); ok {
-					fallback = true
-					confirmedV2 = confirmedV2 || fallbackErr.confirmedV2
-					if fallbackErr.transportOK && endpoint.URL.Scheme == "https" {
-						confirmedTLSRegistries[endpoint.URL.Host] = struct{}{}
-					}
-					err = fallbackErr.err
-				}
 			}
-			if fallback {
-				if _, ok := err.(ErrNoSupport); !ok {
-					// Because we found an error that's not ErrNoSupport, discard all subsequent ErrNoSupport errors.
-					discardNoSupportErrors = true
-					// append subsequent errors
-					lastErr = err
-				} else if !discardNoSupportErrors {
-					// Save the ErrNoSupport error, because it's either the first error or all encountered errors
-					// were also ErrNoSupport errors.
-					// append subsequent errors
-					lastErr = err
+			if fallbackErr, ok := err.(fallbackError); ok {
+				confirmedV2 = confirmedV2 || fallbackErr.confirmedV2
+				if fallbackErr.transportOK && endpoint.URL.Scheme == "https" {
+					confirmedTLSRegistries[endpoint.URL.Host] = struct{}{}
 				}
-				logrus.Errorf("Attempting next endpoint for pull after error: %v", err)
-				continue
+				err = fallbackErr.err
 			}
-			logrus.Errorf("Not continuing with pull after error: %v", err)
-			return TranslatePullError(err, ref)
+			if _, ok := err.(ErrNoSupport); !ok {
+				// Because we found an error that's not ErrNoSupport, discard all subsequent ErrNoSupport errors.
+				discardNoSupportErrors = true
+				// append subsequent errors
+				lastErr = err
+			} else if !discardNoSupportErrors {
+				// Save the ErrNoSupport error, because it's either the first error or all encountered errors
+				// were also ErrNoSupport errors.
+				// append subsequent errors
+				lastErr = err
+			}
+			logrus.Errorf("Attempting next endpoint for pull after error: %v", err)
+			continue
```

Comment:
I made another attempt at simplifying this logic, this time hopefully without breaking anything. Let me know if it looks right to you and if it's an improvement, if not, I can take it out so it doesn't delay this PR.